### PR TITLE
[Caffe2] Fix issues in IDEEP fallback ops and enable Detectron on IDEEP device

### PR DESCRIPTION
**Pull request by gujinghui**
Fix issues in IDEEP fallback ops and enable Detectron on IDEEP device

Fix the correctness issue in fallback op implementation when the fallback op is in-place and its upstream op is with IDEEP device. The reorder would happen incorrectly from the same buffer then.
Support Mask-RCNN model from the Detectron, added all the fallbacks needed, e.g. Python op and other related ops. And also support feeding blobs with integer types using TensorCPU.

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai @jgong5
Pls help review.

**Review by yinghai — COMMENTED** [review]
Hi @gujinghui, what issue does this PR fix and could you outline what the fix is? 

Also please clang-format your code.

**Comment by jgong5** [comment]
@yinghai Below are the fixes and improvements from this PR:
1. Fix the correctness issue in fallback op implementation when the fallback op is in-place and its upstream op is with IDEEP device. The reorder would happen incorrectly from the same buffer then.
2. Support Mask-RCNN model from the Detectron, added all the fallbacks needed, e.g. Python op and other related ops.

File: caffe2/python/pybind_state.cc [hunk 1]
```diff
@@ -229,6 +234,12 @@ class GetPythonGradient : public GradientMakerBase {
   }
 };
 
+#ifdef CAFFE2_USE_IDEEP
+REGISTER_IDEEP_OPERATOR(
+   Python,
+   IDEEPFallbackOp<PythonOp<CPUContext, false>>);
+#endif
+
```
Comment (jgong5) [inline caffe2/python/pybind_state.cc]: Better to place this fallback declaration in pybind_state_ideep.cc?

**Comment by gujinghui** [comment]
@yinghai 

Added some comments and clang-format the codes.
Pls review.

**Comment by gujinghui** [comment]
@yinghai  pls help review and merge.

File: caffe2/ideep/operators/operator_fallback_ideep.h [hunk 2]
```diff
@@ -99,42 +101,48 @@ class IDEEPFallbackOp final : public IDEEPOperator {
...
 
-      if (src.template IsType<float>()) {
-        Blob* dst = OperatorBase::OutputBlob(i);
-        if (!dst->template IsType<itensor>()) {
+      if (src.template IsType<float>() && !SkipOutputCopy::Contains(i) &&
+          base_op_->type() != "Python") {
+        Blob *dst = OperatorBase::OutputBlob(i);
+        // The output tensor must be ideep tensor with public format.
+        // If reusing ideep tensor with non-public format, the tensor buffer
+        // will be wrong interpreted.
```
Comment (yinghai) [inline caffe2/ideep/operators/operator_fallback_ideep.h]: will be interpreted incorrectly.

File: caffe2/ideep/operators/operator_fallback_ideep.h [hunk 3]
```diff
@@ -99,42 +101,48 @@ class IDEEPFallbackOp final : public IDEEPOperator {
...
+        if (!dst->template IsType<itensor>() ||
+            !dst->template Get<itensor>().is_public_format()) {
           dst->Reset(new itensor());
         }
```
Comment (yinghai) [inline caffe2/ideep/operators/operator_fallback_ideep.h]: What's the `else` clause? We just go ahead assume the blob contains `itensor` if the condition is not met? I think this condition should be lifted out.

File: caffe2/python/pybind_state_ideep.cc [hunk 4]
```diff
@@ -146,7 +143,20 @@ class IDeepFeeder : public BlobFeederBase {
...
+        const auto npy_type = PyArray_TYPE(array);
+        const TypeMeta& meta = NumpyTypeToCaffe(npy_type);
+        // TODO: if necessary, use dispatcher.
+        if (meta.Match<float>()) {
+          FeedTensor(option, original_array, blob->GetMutable<itensor>());
+        } else {
+          DeviceOption cpu_option(option);
+          cpu_option.set_device_type(DeviceType::CPU);
+          TensorFeeder<CPUContext> cpu_tensor_feeder;
+          cpu_tensor_feeder.FeedTensor(cpu_option, original_array, blob->GetMutable<TensorCPU>());
```
Comment (yinghai) [inline caffe2/python/pybind_state_ideep.cc]: clang-format?

**Review by yinghai — CHANGES_REQUESTED** [review]

File: caffe2/ideep/operators/operator_fallback_ideep.h (hunk 2 above)
Comment (gujinghui) [reply caffe2/ideep/operators/operator_fallback_ideep.h]: fixed.

File: caffe2/python/pybind_state.cc (hunk 1 above)
Comment (gujinghui) [reply caffe2/python/pybind_state.cc]: fixed

File: caffe2/ideep/operators/operator_fallback_ideep.h (hunk 3 above)
Comment (gujinghui) [reply caffe2/ideep/operators/operator_fallback_ideep.h]: if it's not an itensor, the tensor will not be reused. The GetMutable<itensor> in line \#123 will re-create a new itensor for us.

File: caffe2/python/pybind_state_ideep.cc (hunk 4 above)
Comment (gujinghui) [reply caffe2/python/pybind_state_ideep.cc]: fixed

**Comment by gujinghui** [comment]
@yinghai  rebased and fixed. pls help review.

**Comment by yinghai** [comment]
Still conflict with master?

File: caffe2/python/pybind_state_ideep.cc [hunk 5]
```diff
@@ -105,53 +103,63 @@ class IDeepFeeder : public BlobFeederBase {
...
+      } else {
+        DeviceOption cpu_option(option);
+        cpu_option.set_device_type(DeviceType::CPU);
+        TensorFeeder<CPUContext> cpu_tensor_feeder;
+        cpu_tensor_feeder.FeedTensor(cpu_option, original_array,
+                                     blob->GetMutable<TensorCPU>());
       }
-   }
+    } catch (ideep::error &e) {
+      VLOG(1) << "IDEEP error: " << e.message;
```
Comment (yinghai) [inline caffe2/python/pybind_state_ideep.cc]: `LOG(ERROR)`?

Comment (gujinghui) [reply caffe2/python/pybind_state_ideep.cc]: Fixed

**Comment by yinghai** [comment]
Please rebase as the failure tests are blocking merge. And do you have any tests for the in-place fallback op?

**Comment by gujinghui** [comment]
rebased & added test case. @yinghai

**Comment by jgong5** [comment]
@yinghai Any comments on the updates? OK to merge?

**Comment by jgong5** [comment]
@yinghai Hold on. We need to further polish the fix.

**Comment by gujinghui** [comment]
Move to other PR

**Pull request closed without merging**